### PR TITLE
Record workers as maps

### PR DESCRIPTION
**Pull request by mmmries**
As I was working on #6 I was trying to support killing a worker that had been sent to one of several different delegate pools. It was going to be a lot easier if I could add some optional keys to the worker metadata (ie either have a `task_pid` or a `delegated_to`, but not both). So I decided to create this PR as a "pre-factor".

This PR introduces the change to store workers as maps (either a `local_worker` or a `delegated_worker` map). I also added some typespecs which I found helpful to keep the shape of the state in my head.

As I was making that change I also realized that I could unify the way we kill processes for both local and delegated workers by just pattern matching on whether they had a `task_pid` or a `delegated_to` key. And for the delegated tasks I switched to killing them by their `ref`. This avoids the weird edge cases where we end up killing tasks in the root pool that were actually started through some other child pool. Since `ref`s are guaranteed to be unique by the VM we can treat them as fully unique and just kill the one task we're interested in without having to iterate through all of the workers.

And lastly I switched the process of killing a task so that it sends the kill signal to the task and let's it be killed via the normal process, so we don't have to do special case un-linking and freeing commitments. This also solves an edge case where a task had been delegated through multiple pools and you want all of them to be notified of that process having crashed.

### CODE REVIEW

**Comment by mmmries** [comment]
One other note, there are some more typespecs and docs which would be nice to add, but I'm including those in a follow-up PR. I was on a flight and got like 3 PRs written so I'm trying to open them one-at-a-time to keep them smaller and easier to review.

File: lib/handler/pool/state.ex [hunk 1]
```diff
@@ -14,15 +14,23 @@ defmodule Handler.Pool.State do
...
-              task_name :: String.t()
-            }
+            reference() => worker()
           },
           pool: %Handler.Pool{}
         }
+  @type worker :: local_worker() | delegated_worker()
+  @type local_worker :: %{
+          bytes_committed: non_neg_integer(),
+          from_pid: pid(),
```
Comment (ashton314) [inline lib/handler/pool/state.ex:24]: Bit of a nit-pick, but it might be nice to reorder these fields so that it's more immediately apparent where the difference is. E.g.

```elixir
@type worker :: local_worker() | delegated_worker()
@type local_worker :: %{
        bytes_committed: non_neg_integer(),
        from_pid: pid(),
        task_name: String.t() | nil,
        task_pid: pid()
      }
@type delegated_worker :: %{
        bytes_committed: non_neg_integer(),
        from_pid: pid(),
        task_name: String.t() | nil,
        delegated_to: Pool.pool()
      }
```

File: lib/handler/pool/state.ex [hunk 2]
```diff
@@ -54,59 +62,52 @@ defmodule Handler.Pool.State do
...
-      when not is_nil(pool) do
-    if has_worker_named(state, task_name) do
-      {:ok, number_killed} = Pool.kill(pool, task_name)
-      {:ok, number_killed, state}
-    else
-      {:ok, 0, state}
-    end
+  @spec kill_worker(t(), String.t()) :: {:ok, t(), non_neg_integer()}
+  def kill_worker(state, task_name) do
+    Enum.reduce(state.workers, {:ok, state, 0}, fn
```
Comment (ashton314) [inline lib/handler/pool/state.ex:75]: Is this function meant to be part of the module's API? Or is it public incidentally for testing purposes? If it's the former, would be nice to have a function doc :+1:

File: lib/handler/pool/state.ex [hunk 3]
```diff
@@ -54,59 +62,52 @@ defmodule Handler.Pool.State do
...
+      {_ref, %{task_pid: task_pid, task_name: ^task_name}}, {:ok, state, number_killed} ->
+        :ok = Process.send(task_pid, {Pool, :user_killed}, [])
+        {:ok, state, number_killed + 1}
```
Comment (ashton314) [inline lib/handler/pool/state.ex:78]: Ah, reduce... king of the list processing higher-order functions. Nice use here---really clear :+1:

File: lib/handler/pool/state.ex [hunk 4]
```diff
@@ -54,59 +62,52 @@ defmodule Handler.Pool.State do
...
+    case Map.get(workers, ref) do
+      %{delegated_to: pool} ->
+        Pool.kill_by_ref(pool, ref)
 
-    {:ok, number_killed, state}
+      %{task_pid: pid} ->
+        :ok = Process.send(pid, {Pool, :user_killed}, [])
+        :ok
+    end
   end
```
Comment (ashton314) [inline lib/handler/pool/state.ex:109]: Under what circumstances would `Process.send/3` fail, and why would blowing up be the right thing to do?

**Review by ashton314 — APPROVED** [review]
This pre-factor looks good! I've left a few comments, but nothing seems to be particularly *blocking*.

File: lib/handler/pool/state.ex (hunk 1 above)
Comment (mmmries) [reply lib/handler/pool/state.ex:24]: 👍 I like that idea

File: lib/handler/pool/state.ex (hunk 2 above)
Comment (mmmries) [reply lib/handler/pool/state.ex:75]: It is meant to be public (it was added back in #4). I'm planning to add docs for this in one of my follow-on PRs

File: lib/handler/pool/state.ex (hunk 4 above)
Comment (mmmries) [reply lib/handler/pool/state.ex:109]: That's a great question. When as I was writing this the only use-case I could imagine where it would fail would be a race condition where a process completes and exits as we are in the process of trying to kill it. I think that race condition will be pretty rare since these `kill` operators are pretty fast so it would be a narrow window of time.

But, message delivery is not explicitly guaranteed by the VM and I'm sure it can fail for other reasons such as network interruptions. I wasn't sure of the best way to handle those failures, so I decided that having the process crash was the the safest option given that this was a pretty unknown space of potential problems.

TL/DR; If I know which errors to account for I'll try to recover, but when I don't know how to recover I follow the "let it crash" philosophy because that will force the process to "turn it off and turn it back on" to get a nice clean state

File: test/handler/pool_test.exs [hunk 5]
```diff
@@ -283,18 +275,33 @@ defmodule Handler.PoolTest do
...
+      :sys.get_state(composed)
 
       {:ok, 1} = Pool.kill(composed, task_name)
+      assert {:error, exception} = Pool.await(ref)
+
+      assert exception == %Handler.ProcessExit{
+               reason: :user_killed,
+               message: "User killed the process"
+             }
+
```
Comment (marcelloma) [inline test/handler/pool_test.exs:292]: We did similar changes to this test(it was very flawed before) but one thing I don't see on this PR is an assertion that the task did in fact die. One of the issues that was happening when using this was the because Task.async had it's exited trapped, it would never really die, but we would return to the client the proper ProcessExit message.

Can you add refute Process.alive?(pid) here?

File: lib/handler/pool/state.ex [hunk 6]
```diff
@@ -54,59 +62,52 @@ defmodule Handler.Pool.State do
...
-      {:ok, number_killed} = Pool.kill(pool, task_name)
-      {:ok, number_killed, state}
-    else
-      {:ok, 0, state}
-    end
+  @spec kill_worker(t(), String.t()) :: {:ok, t(), non_neg_integer()}
+  def kill_worker(state, task_name) do
+    Enum.reduce(state.workers, {:ok, state, 0}, fn
+      {_ref, %{task_pid: task_pid, task_name: ^task_name}}, {:ok, state, number_killed} ->
+        :ok = Process.send(task_pid, {Pool, :user_killed}, [])
```
Comment (marcelloma) [inline lib/handler/pool/state.ex]: So the pool will now rely on the handler to commit seppuku, instead of the pool killing the wrapper task directly. When we fixed this yesterday, we used Task.shutdown(w/ brutal_kill)  from the pool instead, and removed all mentions of user killing from Handler.run. I think the end result is the same, do you think using Task.shutdown from the pool would have any drawbacks in comparison to this approach? What I don't like about this is that Handler has to acknowledge that it may belong to a pool and that it's pool may ask it do kill itself.

**Comment by marcelloma** [comment]
We solved some of the issues this PR solves on this other PR: https://github.com/SpiffInc/handler/pull/10

**Comment by mmmries** [comment]
Oh that's super interesting that you and Luís were looking into this too. I'll take a look at #10 now and see if using `Task.shutdown` is a better option 👍

**Comment by luismaldonadov** [comment]
I will be closing PR #7  as this is a confirmed fix for the problem we were having.

**Comment by mmmries** [comment]
@marcelloma @luismaldonadov I tried to switch to using `Task.shutdown`, but when you use that function it flushes out all messages from the process that is being killed, this means we have to include code to update the state and cleanup the resource commitments at the point where we call `Task.shutdown`. After playing with the code a bit, I prefer the method of sending an exit signal and letting the results from the task be sent back to the pool process as per normal. This way we re-use the existing code that watching for responses and/or `:DOWN` messages from our tasks and does all the resource cleanup in the normal way

Were there other reasons that you preferred `Task.shutdown` for this use-case?

**Comment by marcelloma** [comment]
Yes, by removing signal to handler you would need to cleanup the commitments and send_response from the pool, right after Task.shutdown is called. While I do prefer it because this way Handler does not need to acknowledge that Handler.Pool may send it a kill message, I get your point about not doing the same work twice(cleanup/send_response).

**Review by marcelloma — APPROVED** [review]

**Comment by mmmries** [comment]
@marcelloma I've updated this PR to make killing a task synchronous (the task process is guaranteed to be killed by them the client receives its response). This should facilitate us adding a `Pool.flush` function for killing all of the current work in a pool. I'll tackle that API in a separate PR (along with some other follow-on PRs that I have already prepared) and then we can release a version update to hex with these changes

**Review by marcelloma — APPROVED** [review]
